File: Bridge/System/Guid.cs
using Bridge;

using System.Runtime.CompilerServices;

namespace System
{
    /// <summary>
    /// The Guid data type which is mapped to the string type in Javascript.
    /// </summary>
    [Immutable]
    public struct Guid : IEquatable<Guid>, IComparable<Guid>, IFormattable
    {
        private const string error1 = "Byte array for GUID must be exactly {0} bytes long";

        private static readonly Bridge.Text.RegularExpressions.Regex Valid = new Bridge.Text.RegularExpressions.Regex("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", "i");
        private static readonly Bridge.Text.RegularExpressions.Regex Split = new Bridge.Text.RegularExpressions.Regex("^(.{8})(.{4})(.{4})(.{4})(.{12})$");
        private static readonly Bridge.Text.RegularExpressions.Regex NonFormat = new Bridge.Text.RegularExpressions.Regex("^[{(]?([0-9a-f]{8})-?([0-9a-f]{4})-?([0-9a-f]{4})-?([0-9a-f]{4})-?([0-9a-f]{12})[)}]?$", "i");
        private static readonly Bridge.Text.RegularExpressions.Regex Replace = new Bridge.Text.RegularExpressions.Regex("-", "g");
        private static readonly Random Rnd = new Random();

        /// <summary>
        /// A read-only instance of the Guid structure whose value is all zeros.
        /// </summary>
        public static readonly Guid Empty = new Guid();

        private int _a;
        private short _b;
        private short _c;
        private byte _d;
        private byte _e;
        private byte _f;
        private byte _g;
        private byte _h;
        private byte _i;
        private byte _j;
        private byte _k;

        /// <summary>
        /// Initializes a new instance of the Guid structure by using the value represented by the specified string.
        /// </summary>
        /// <param name="uuid">A string that contains a GUID</param>
        public Guid(string uuid)
        {
            this = new Guid();

            ParseInternal(uuid, null, true);
        }

        /// <summary>
        /// Initializes a new instance of the Guid structure by using the specified array of bytes.
        /// </summary>
        /// <param name="b">A 16-element byte array containing values with which to initialize the GUID.</param>
        public Guid(byte[] b)
        {
            if (b == null)
            {
                throw new ArgumentNullException("b");
            }

            if (b.Length != 16)
            {
                throw new ArgumentException(string.Format(error1, 16));
            }

            _a = ((int)b[3] << 24) | ((int)b[2] << 16) | ((int)b[1] << 8) | b[0];
            _b = (short)(((int)b[5] << 8) | b[4]);
            _c = (short)(((int)b[7] << 8) | b[6]);
            _d = b[8];
            _e = b[9];
            _f = b[10];
            _g = b[11];
            _h = b[12];
            _i = b[13];
            _j = b[14];
            _k = b[15];
        }

        /// <summary>
        /// Initializes a new instance of the Guid structure by using the specified unsigned integers and bytes.
        /// </summary>
        /// <param name="a">The first 4 bytes of the GUID.</param>
        /// <param name="b">The next 2 bytes of the GUID.</param>
        /// <param name="c">The next 2 bytes of the GUID.</param>
        /// <param name="d">The next byte of the GUID.</param>
        /// <param name="e">The next byte of the GUID.</param>
        /// <param name="f">The next byte of the GUID.</param>
        /// <param name="g">The next byte of the GUID.</param>
        /// <param name="h">The next byte of the GUID.</param>
        /// <param name="i">The next byte of the GUID.</param>
        /// <param name="j">The next byte of the GUID.</param>
        /// <param name="k">The next byte of the GUID.</param>
        public Guid(uint a, ushort b, ushort c, byte d, byte e, byte f, byte g, byte h, byte i, byte j, byte k)
        {
            _a = (int)a;
            _b = (short)b;
            _c = (short)c;
            _d = d;
            _e = e;
            _f = f;
            _g = g;
            _h = h;
            _i = i;
            _j = j;
            _k = k;
        }

        /// <summary>
        /// Initializes a new instance of the Guid structure by using the specified integers and byte array.
        /// </summary>
        /// <param name="a">The first 4 bytes of the GUID.</param>
        /// <param name="b">The next 2 bytes of the GUID.</param>
        /// <param name="c">The next 2 bytes of the GUID.</param>
        /// <param name="d">The remaining 8 bytes of the GUID.</param>
        public Guid(int a, short b, short c, byte[] d)
        {
            if (d == null)
            {
                throw new ArgumentNullException("d");
            }

            if (d.Length != 8)
            {
                throw new ArgumentException(string.Format(error1, 8));
            }

            _a = a;
            _b = b;
            _c = c;
            _d = d[0];
            _e = d[1];
            _f = d[2];
            _g = d[3];
            _h = d[4];
            _i = d[5];
            _j = d[6];
            _k = d[7];
        }

        /// <summary>
        /// Initializes a new instance of the Guid structure by using the specified integers and bytes.
        /// </summary>
        /// <param name="a">The first 4 bytes of the GUID.</param>
        /// <param name="b">The next 2 bytes of the GUID.</param>
        /// <param name="c">The next 2 bytes of the GUID.</param>
        /// <param name="d">The next byte of the GUID.</param>
        /// <param name="e">The next byte of the GUID.</param>
        /// <param name="f">The next byte of the GUID.</param>
        /// <param name="g">The next byte of the GUID.</param>
        /// <param name="h">The next byte of the GUID.</param>
        /// <param name="i">The next byte of the GUID.</param>
        /// <param name="j">The next byte of the GUID.</param>
        /// <param name="k">The next byte of the GUID.</param>
        public Guid(int a, short b, short c, byte d, byte e, byte f, byte g, byte h, byte i, byte j, byte k)
        {
            _a = a;
            _b = b;
            _c = c;
            _d = d;
            _e = e;
            _f = f;
            _g = g;
            _h = h;
            _i = i;
            _j = j;
            _k = k;
        }

        /// <summary>
        /// Returns a value indicating whether this instance and a specified Guid object represent the same value.
        /// </summary>
        /// <param name="o">An object to compare to this instance.</param>
        /// <returns>true if o is equal to this instance; otherwise, false.</returns>
        public bool Equals(Guid o)
        {
            if ((this._a != o._a) || (this._b != o._b) || (this._c != o._c) || (this._d != o._d)
                || (this._e != o._e) || (this._f != o._f) || (this._g != o._g) || (this._h != o._h)
                || (this._i != o._i) || (this._j != o._j) || (this._k != o._k))
            {
                return false;
            }

            return true;
        }

        /// <summary>
        /// Compares this instance to a specified Guid object and returns an indication of their relative values.
        /// </summary>
        /// <param name="value">An object to compare to this instance.</param>
        /// <returns>A signed number indicating the relative values of this instance and value.</returns>
        public int CompareTo(Guid value)
        {
            return this.ToString().CompareTo(value.ToString());
        }

        /// <summary>
        /// Returns a string representation of the value of this instance in registry format.
        /// </summary>
        /// <returns>The value of this Guid, formatted by using the "D" format specifier.</returns>
        public override string ToString()
        {
            return Format(null);
        }

        /// <summary>
        /// Returns a string representation of the value of this Guid instance, according to the provided format specifier.
        /// </summary>
        /// <param name="format">A single format specifier that indicates how to format the value of this Guid. The format parameter can be "N", "D", "B", "P". If format is null or an empty string (""), "D" is used.</param>
        /// <returns>The value of this Guid, represented as a series of lowercase hexadecimal digits in the specified format.</returns>
        public string ToString(string format)
        {
            return Format(format);
        }

        /// <summary>
        /// Returns a string representation of the value of this instance of the Guid class, according to the provided format specifier and culture-specific format information.
        /// </summary>
        /// <param name="format">A single format specifier that indicates how to format the value of this Guid. The format parameter can be "N", "D", "B", "P". If format is null or an empty string (""), "D" is used.</param>
        /// <param name="formatProvider">(Reserved) An object that supplies culture-specific formatting information.</param>
        /// <returns>The value of this Guid, represented as a series of lowercase hexadecimal digits in the specified format.</returns>
        public string ToString(string format, IFormatProvider formatProvider)
        {
            return Format(format);
        }

        /// <summary>
        /// Returns a 16-element byte array that contains the value of this instance.
        /// </summary>
        /// <returns>A 16-element byte array.</returns>
        public byte[] ToByteArray()
        {
            byte[] g = new byte[16];

            g[0] = (byte)(_a);
            g[1] = (byte)(_a >> 8);
            g[2] = (byte)(_a >> 16);
            g[3] = (byte)(_a >> 24);
            g[4] = (byte)(_b);
            g[5] = (byte)(_b >> 8);
            g[6] = (byte)(_c);
            g[7] = (byte)(_c >> 8);
            g[8] = _d;
            g[9] = _e;
            g[10] = _f;
            g[11] = _g;
            g[12] = _h;
            g[13] = _i;
            g[14] = _j;
            g[15] = _k;

            return g;
        }

        /// <summary>
        /// Converts the string representation of a GUID to the equivalent Guid structure.
        /// </summary>
        /// <param name="input">The string to convert.</param>
        /// <returns>A structure that contains the value that was parsed.</returns>
        public static Guid Parse(string input)
        {
            return ParseExact(input, null);
        }

        /// <summary>
        /// Converts the string representation of a GUID to the equivalent Guid structure, provided that the string is in the specified format.
        /// </summary>
        /// <param name="input">The GUID to convert.</param>
        /// <param name="format">One of the following specifiers that indicates the exact format to use when interpreting input: "N", "D", "B", "P".</param>
        /// <returns></returns>
        public static Guid ParseExact(string input, string format)
        {
            var r = new Guid();
            r.ParseInternal(input, format, true);
            return r;
        }

        /// <summary>
        /// Converts the string representation of a GUID to the equivalent Guid structure.
        /// </summary>
        /// <param name="input">The GUID to convert.</param>
        /// <param name="result">The structure that will contain the parsed value. If the method returns true, result contains a valid Guid. If the method returns false, result equals Guid.Empty.</param>
        /// <returns></returns>
        public static bool TryParse(string input, out Guid result)
        {
            return TryParseExact(input, null, out result);
        }

        /// <summary>
        /// Converts the string representation of a GUID to the equivalent Guid structure, provided that the string is in the specified format.
        /// </summary>
        /// <param name="input">The GUID to convert.</param>
        /// <param name="format">One of the following specifiers that indicates the exact format to use when interpreting input: "N", "D", "B", "P".</param>
        /// <param name="result">The structure that will contain the parsed value. If the method returns true, result contains a valid Guid. If the method returns false, result equals Guid.Empty.</param>
        /// <returns></returns>
        public static bool TryParseExact(string input, string format, out Guid result)
        {
            result = new Guid();
            return result.ParseInternal(input, format, false);
        }

        /// <summary>
        /// Initializes a new instance of the Guid structure.
        /// </summary>
        /// <returns>A new GUID object.</returns>
        public static Guid NewGuid()
        {
            var a = new byte[16];

            Rnd.NextBytes(a);

            a[7] = (byte)(a[7] & 0x0f | 0x40);
            a[8] = (byte)(a[8] & 0xbf | 0x80);

            return new Guid(a);
        }

        /// <summary>
        /// Indicates whether the values of two specified Guid objects are equal.
        /// </summary>
        /// <param name="a">The first object to compare.</param>
        /// <param name="b">The second object to compare.</param>
        /// <returns>true if a and b are equal; otherwise, false.</returns>
        public static bool operator ==(Guid a, Guid b)
        {
            if (Object.ReferenceEquals(a, null))
            {
                return Object.ReferenceEquals(b, null);
            }

            return a.Equals(b);
        }

        /// <summary>
        /// Indicates whether the values of two specified Guid objects are not equal.
        /// </summary>
        /// <param name="a">The first object to compare.</param>
        /// <param name="b">The second object to compare.</param>
        /// <returns>true if a and b are not equal; otherwise, false.</returns>
        public static bool operator !=(Guid a, Guid b)
        {
            return !(a == b);
        }

        private bool ParseInternal(string input, string format, bool check)
        {
            string r = null;

            if (string.IsNullOrEmpty(input))
            {
                throw new System.ArgumentNullException("input");
            }

            if (string.IsNullOrEmpty(format))
            {
                var m = NonFormat.Exec(input);

                if (m != null)
                {
                    r = m.Slice(1).Join("-").ToLower();
                }
            }
            else
            {
                format = format.ToUpper();

                var p = false;

                if (format == "N")
                {
                    var m = Split.Exec(input);

                    if (m != null)
                    {
                        p = true;
                        input = m.Slice(1).Join("-");
                    }
                }
                else if (format == "B" || format == "P")
                {
                    var b = format == "B" ? new[] { '{', '}' } : new[] { '(', ')' };

                    if ((input[0] == b[0]) && (input[input.Length - 1] == b[1]))
                    {
                        p = true;
                        input = input.Substr(1, input.Length - 2);
                    }
                } else
                {
                    p = true;
                }

                if (p && input.Match(Valid) != null)
                {
                    r = input.ToLower();
                }
            }

            if (r != null)
            {
                FromString(r);
                return true;
            }

            if (check)
            {
                throw new System.FormatException("input is not in a recognized format");
            }

            return false;
        }

        private string Format(string format)
        {
            var s = ((uint)_a).ToString("x8") + ((ushort)_b).ToString("x4") + ((ushort)_c).ToString("x4");
            s = s + (new[] { _d, _e, _f, _g, _h, _i, _j, _k }).Map(MakeBinary).Join("");

            s = Split.Exec(s).Slice(1).Join("-");

            switch (format)
            {
                case "n":
                case "N":
                    return s.Replace(Replace, "");
                case "b":
                case "B":
                    return '{' + s + '}';
                case "p":
                case "P":
                    return '(' + s + ')';
                default:
                    return s;
            }
        }

        private static string MakeBinary(byte x)
        {
            return (x & 0xff).ToString("x2");
        }

        private void FromString(string s)
        {
            if (string.IsNullOrEmpty(s))
            {
                return;
            }

            s = s.Replace(Replace, "");

            var r = new byte[8];

            _a = (int)uint.Parse(s.Substring(0, 8), 16);
            _b = (short)ushort.Parse(s.Substring(8, 4), 16);
            _c = (short)ushort.Parse(s.Substring(12, 4), 16);
            for (var i = 8; i < 16; i++)
            {
                r[i - 8] = byte.Parse(s.Substring(i * 2, 2), 16);
            }

            _d = r[0];
            _e = r[1];
            _f = r[2];
            _g = r[3];
            _h = r[4];
            _i = r[5];
            _j = r[6];
            _k = r[7];
        }
    }
}

File: Bridge/Resources/.generated/system/guid.js
    Bridge.define("System.Guid", {
        inherits: function () { return [System.IEquatable$1(System.Guid),System.IComparable$1(System.Guid),System.IFormattable]; },
        $kind: "struct",
        statics: {
            error1: "Byte array for GUID must be exactly {0} bytes long",
            valid: null,
            split: null,
            nonFormat: null,
            replace: null,
            rnd: null,
            config: {
                init: function () {
                    this.valid = new RegExp("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", "i");
                    this.split = new RegExp("^(.{8})(.{4})(.{4})(.{4})(.{12})$");
                    this.nonFormat = new RegExp("^[{(]?([0-9a-f]{8})-?([0-9a-f]{4})-?([0-9a-f]{4})-?([0-9a-f]{4})-?([0-9a-f]{12})[)}]?$", "i");
                    this.replace = new RegExp("-", "g");
                    this.rnd = new System.Random.ctor();
                    this.empty = new System.Guid.ctor();
                }
            },
            parse: function (input) {
                return System.Guid.parseExact(input, null);
            },
            parseExact: function (input, format) {
                var r = new System.Guid.ctor();
                r.parseInternal(input, format, true);
                return r;
            },
            tryParse: function (input, result) {
                return System.Guid.tryParseExact(input, null, result);
            },
            tryParseExact: function (input, format, result) {
                result.v = new System.Guid.ctor();
                return result.v.parseInternal(input, format, false);
            },
            newGuid: function () {
                var a = System.Array.init(16, 0);

                System.Guid.rnd.nextBytes(a);

                a[7] = ((a[7] & 15 | 64)) & 255;
                a[8] = ((a[8] & 191 | 128)) & 255;

                return new System.Guid.$ctor1(a);
            },
            makeBinary: function (x) {
                return System.Int32.format((x & 255), "x2");
            },
            op_Equality: function (a, b) {
                if (Bridge.referenceEquals(a, null)) {
                    return Bridge.referenceEquals(b, null);
                }

                return a.equalsT(b);
            },
            op_Inequality: function (a, b) {
                return !(System.Guid.op_Equality(a, b));
            },
            getDefaultValue: function () { return new System.Guid(); }
        },
        _a: 0,
        _b: 0,
        _c: 0,
        _d: 0,
        _e: 0,
        _f: 0,
        _g: 0,
        _h: 0,
        _i: 0,
        _j: 0,
        _k: 0,
        config: {
            alias: [
            "equalsT", "System$IEquatable$1$System$Guid$equalsT",
            "compareTo", "System$IComparable$1$System$Guid$compareTo",
            "format", "System$IFormattable$format"
            ]
        },
        $ctor4: function (uuid) {
            this.$initialize();
            (new System.Guid.ctor()).$clone(this);

            this.parseInternal(uuid, null, true);
        },
        $ctor1: function (b) {
            this.$initialize();
            if (b == null) {
                throw new System.ArgumentNullException("b");
            }

            if (b.length !== 16) {
                throw new System.ArgumentException(System.String.format(System.Guid.error1, 16));
            }

            this._a = (b[3] << 24) | (b[2] << 16) | (b[1] << 8) | b[0];
            this._b = Bridge.Int.sxs((((b[5] << 8) | b[4])) & 65535);
            this._c = Bridge.Int.sxs((((b[7] << 8) | b[6])) & 65535);
            this._d = b[8];
            this._e = b[9];
            this._f = b[10];
            this._g = b[11];
            this._h = b[12];
            this._i = b[13];
            this._j = b[14];
            this._k = b[15];
        },
        $ctor5: function (a, b, c, d, e, f, g, h, i, j, k) {
            this.$initialize();
            this._a = a | 0;
            this._b = Bridge.Int.sxs(b & 65535);
            this._c = Bridge.Int.sxs(c & 65535);
            this._d = d;
            this._e = e;
            this._f = f;
            this._g = g;
            this._h = h;
            this._i = i;
            this._j = j;
            this._k = k;
        },
        $ctor3: function (a, b, c, d) {
            this.$initialize();
            if (d == null) {
                throw new System.ArgumentNullException("d");
            }

            if (d.length !== 8) {
                throw new System.ArgumentException(System.String.format(System.Guid.error1, 8));
            }

            this._a = a;
            this._b = b;
            this._c = c;
            this._d = d[0];
            this._e = d[1];
            this._f = d[2];
            this._g = d[3];
            this._h = d[4];
            this._i = d[5];
            this._j = d[6];
            this._k = d[7];
        },
        $ctor2: function (a, b, c, d, e, f, g, h, i, j, k) {
            this.$initialize();
            this._a = a;
            this._b = b;
            this._c = c;
            this._d = d;
            this._e = e;
            this._f = f;
            this._g = g;
            this._h = h;
            this._i = i;
            this._j = j;
            this._k = k;
        },
        ctor: function () {
            this.$initialize();
        },
        equalsT: function (o) {
            if ((this._a !== o._a) || (this._b !== o._b) || (this._c !== o._c) || (this._d !== o._d) || (this._e !== o._e) || (this._f !== o._f) || (this._g !== o._g) || (this._h !== o._h) || (this._i !== o._i) || (this._j !== o._j) || (this._k !== o._k)) {
                return false;
            }

            return true;
        },
        compareTo: function (value) {
            return System.String.compare(this.toString(), value.toString());
        },
        toString: function () {
            return this.format$1(null);
        },
        toString$1: function (format) {
            return this.format$1(format);
        },
        format: function (format, formatProvider) {
            return this.format$1(format);
        },
        toByteArray: function () {
            var g = System.Array.init(16, 0);

            g[0] = (this._a) & 255;
            g[1] = ((this._a >> 8)) & 255;
            g[2] = ((this._a >> 16)) & 255;
            g[3] = ((this._a >> 24)) & 255;
            g[4] = (this._b) & 255;
            g[5] = ((this._b >> 8)) & 255;
            g[6] = (this._c) & 255;
            g[7] = ((this._c >> 8)) & 255;
            g[8] = this._d;
            g[9] = this._e;
            g[10] = this._f;
            g[11] = this._g;
            g[12] = this._h;
            g[13] = this._i;
            g[14] = this._j;
            g[15] = this._k;

            return g;
        },
        parseInternal: function (input, format, check) {
            var r = null;

            if (System.String.isNullOrEmpty(input)) {
                throw new System.ArgumentNullException("input");
            }

            if (System.String.isNullOrEmpty(format)) {
                var m = System.Guid.nonFormat.exec(input);

                if (m != null) {
                    r = m.slice(1).join("-").toLowerCase();
                }
            } else {
                format = format.toUpperCase();

                var p = false;

                if (Bridge.referenceEquals(format, "N")) {
                    var m1 = System.Guid.split.exec(input);

                    if (m1 != null) {
                        p = true;
                        input = m1.slice(1).join("-");
                    }
                } else if (Bridge.referenceEquals(format, "B") || Bridge.referenceEquals(format, "P")) {
                    var b = Bridge.referenceEquals(format, "B") ? [123, 125] : [40, 41];

                    if ((input.charCodeAt(0) === b[0]) && (input.charCodeAt(((input.length - 1) | 0)) === b[1])) {
                        p = true;
                        input = input.substr(1, ((input.length - 2) | 0));
                    }
                } else {
                    p = true;
                }

                if (p && input.match(System.Guid.valid) != null) {
                    r = input.toLowerCase();
                }
            }

            if (r != null) {
                this.fromString(r);
                return true;
            }

            if (check) {
                throw new System.FormatException("input is not in a recognized format");
            }

            return false;
        },
        format$1: function (format) {
            var s = System.String.concat(System.String.concat(System.UInt32.format((this._a >>> 0), "x8"), System.UInt16.format((this._b & 65535), "x4")), System.UInt16.format((this._c & 65535), "x4"));
            s = System.String.concat(s, ([this._d, this._e, this._f, this._g, this._h, this._i, this._j, this._k]).map(System.Guid.makeBinary).join(""));

            s = System.Guid.split.exec(s).slice(1).join("-");

            switch (format) {
                case "n": 
                case "N": 
                    return s.replace(System.Guid.replace, "");
                case "b": 
                case "B": 
                    return System.String.concat(System.String.concat(String.fromCharCode(123), s), String.fromCharCode(125));
                case "p": 
                case "P": 
                    return System.String.concat(System.String.concat(String.fromCharCode(40), s), String.fromCharCode(41));
                default: 
                    return s;
            }
        },
        fromString: function (s) {
            if (System.String.isNullOrEmpty(s)) {
                return;
            }

            s = s.replace(System.Guid.replace, "");

            var r = System.Array.init(8, 0);

            this._a = (System.UInt32.parse(s.substr(0, 8), 16)) | 0;
            this._b = Bridge.Int.sxs((System.UInt16.parse(s.substr(8, 4), 16)) & 65535);
            this._c = Bridge.Int.sxs((System.UInt16.parse(s.substr(12, 4), 16)) & 65535);
            for (var i = 8; i < 16; i = (i + 1) | 0) {
                r[((i - 8) | 0)] = System.Byte.parse(s.substr(((i * 2) | 0), 2), 16);
            }

            this._d = r[0];
            this._e = r[1];
            this._f = r[2];
            this._g = r[3];
            this._h = r[4];
            this._i = r[5];
            this._j = r[6];
            this._k = r[7];
        },
        getHashCode: function () {
            var h = Bridge.addHash([1684632903, this._a, this._b, this._c, this._d, this._e, this._f, this._g, this._h, this._i, this._j, this._k]);
            return h;
        },
        $clone: function (to) { return this; }
    });
